core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -366,17 +369,88 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
 
+        final HttpResponse res;
+        req.init(reqCtx);
+        if (needsDirectExecution) {
+            res = serve0(req, serviceCfg, service, reqCtx);
+        } else {
+            res = HttpResponse.of(() -> {
+                return serve0(req, serviceCfg, service, reqCtx);
+            }, serviceEventLoop);
+        }
+
+        // Keep track of the number of unfinished requests and
+        // clean up the request stream when response stream ends.
+        final boolean isTransientService =
+                serviceCfg.service().as(TransientService.class) != null;
+        if (!isTransientService) {
+            gracefulShutdownSupport.inc();
+        }
+        unfinishedRequests.put(req, res);
+
+        if (service.shouldCachePath(routingCtx.path(), routingCtx.query(), routed.route())) {
+            reqCtx.log().whenComplete().thenAccept(log -> {
+                final int statusCode = log.responseHeaders().status().code();
+                if (statusCode >= 200 && statusCode < 400) {
+                    RequestTargetCache.putForServer(req.path(), routingCtx.requestTarget());
+                }
+            });
+        }
+
+        final RequestAndResponseCompleteHandler handler =
+                new RequestAndResponseCompleteHandler(channelEventLoop, ctx, reqCtx, req,
+                                                      isTransientService);
+        req.whenComplete().handle(handler.requestCompleteHandler);
+
+        // A future which is completed when the all response objects are written to channel and
+        // the returned promises are done.
+        final CompletableFuture<Void> resWriteFuture = new CompletableFuture<>();
+        resWriteFuture.handle(handler.responseCompleteHandler);
+
+        // Set the response to the request in order to be able to immediately abort the response
+        // when the peer cancels the stream.
+        req.setResponse(res);
+
+        if (req.isHttp1WebSocket()) {
+            assert responseEncoder instanceof Http1ObjectEncoder;
+            final WebSocketHttp1ResponseSubscriber resSubscriber =
+                    new WebSocketHttp1ResponseSubscriber(ctx, responseEncoder, reqCtx, req, resWriteFuture);
+            res.subscribe(resSubscriber, channelEventLoop, SubscriptionOption.WITH_POOLED_OBJECTS);
+        } else if (reqCtx.exchangeType().isResponseStreaming()) {
+            final AbstractHttpResponseSubscriber resSubscriber =
+                    new HttpResponseSubscriber(ctx, responseEncoder, reqCtx, req, resWriteFuture);
+            res.subscribe(resSubscriber, channelEventLoop, SubscriptionOption.WITH_POOLED_OBJECTS);
+        } else {
+            final AggregatedHttpResponseHandler resHandler =
+                    new AggregatedHttpResponseHandler(ctx, responseEncoder, reqCtx, req, resWriteFuture);
+            res.aggregate(AggregationOptions.usePooledObjects(ctx.alloc(), channelEventLoop))
+               .handle(resHandler);
```

Comment:
Question) Don't we need to use `serviceEventLoop` to subscribe responses? Otherwise, channelEventLoop could be exposed easily by the callbacks of futures.
For example `res.whenConsumed()` will be completed by `channelEventLoop` and then `listener.onReady()` will be called on the same thread. https://github.com/line/armeria/blob/a0a7f468f260b7d4c58dcb6e1fdcfbd1f60c80b2/grpc/src/main/java/com/linecorp/armeria/server/grpc/StreamingServerCall.java#L162-L168

One possible solution is to subscribe to the response with `serviceEventLoop` and switch to `channelEventLoop` before writing to the channel. Otherwise, it is also good to switch to `serviceEventLoop` before invoking `invokeOnReady()` or by specifying `serviceEventLoop` to `thenRun()`

Reply:
> One possible solution is to subscribe to the response with serviceEventLoop and switch to channelEventLoop before writing to the channel.

The problem I had with this was that the boundary between `HttpResponseSubscriber` and `HttpObjectEncoder` wasn't clearly cut when I last checked. (an example is `HttpObjectEncoder#isWritable` where `HttpResponseSubscriber` queries a state from `HttpObjectEncoder` which is a different thread)

This PR isn't perfect like you mentioned, if you feel like this is a blocker we can suspend this PR until the above is fixed.

Reply:
That's a good point. We've implemented the callbacks of `whenConsumed()` with the assumption that they will be executed by the eventLoop. However, the stream message can be subscribed by any executor so we should fix all the callbacks of `whenConsumed()`.

Reply:
I see, let me just delay this PR then 😄